### PR TITLE
feat(ec2): make it possible to override mapPublicIpOnLaunch for public Subnets

### LINKED ISSUE #14194: (ec2): Add option to turn Auto-assign public IPv4 address off for public subnets
I need to be able to create Public Subnets with "Auto-assign public IPv4 address" turned off and I would like to do that using the ec2.Vpc functionality and not go full on manual just because of this tiny detail.

### Use Case

Corporate Regulations that can not be changed.

### Proposed Solution

```
const vpc = new ec2.Vpc(this, "Test123", {
      maxAzs: 3,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: "PublicSubnet",
          subnetType: ec2.SubnetType.PUBLIC,
          mapPublicIpOnLaunch: false
        }
      ],
    });
```

### Other

* [X] :wave: I should be able to implement this feature request

As default can stay as it is, there shouldn't be any impact.

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by sandrom**
This enables to override the Auto Assign Public IP Parameter, so you can also turn it off for public subnets.

Fixes #14194

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by sandrom** [comment]
I am not sure this whole approach of adding subnets to the init of the vpc is even tested at the moment. The current test cases I found do not seem to cover this. Should this be changed for this Pull Request or do we consider this simple enough?

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -958,6 +964,13 @@ export interface SubnetConfiguration {
    * @default false
    */
   readonly reserved?: boolean;
+
+  /**
+   * Controls if a public IP is associated to an instance at launch
+   *
+   * @default true in Subnet.Public, false in Subnet.Private or Subnet.Isolated.
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:971]: Should have a comment to say it ONLY works if subnetType == PUBLIC.

Should have a validation that it's not set to true if subnetType != PUBLIC.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Also needs a test

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts (hunk 1 above)
Comment (sandrom) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts:974]: Sorry I was on leave and did not see it earlier. Added the comments, makes sense. You mentioned something about a test, but I am unsure on how to add that.

**Comment by azizur** [comment]
Hey, @sandrom and @rix0rrr any idea when this could be merged?

My AWS Security hub is showing up as a violation of [AWS Foundational Security Best Practices standard [EC2.15] EC2 subnets should not automatically assign public IP addresses](https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#fsbp-ec2-15)

@sandrom if you are busy, I am happy to continue completing the outstanding actions.

**Comment by sandromlp** [comment]
@azizur it needs a test but not sure how to best test this as the suite only covers parts anyway and this is rather specific. Time was an issue for me but also no reply on my question. If you have more time, feel free! 🙂

**Pull request closed without merging**